R/CompDbSource.R

```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
+        testQuery <- list(
+            fragmentpeakmatchabsolutemassdeviation = "0.001",
+            fragmentpeakmatchrelativemassdeviation = "5",
+            databasesearchrelativemassdeviation = "5",
+            metfragdatabasetype = "PubChem",
+            peakliststring = paste0("177.092320760091_1451.08410644531;",
+                                    "205.086236000061_1983.02587890625;",
+                                    "229.086613972982_1455.84973144531;",
+                                    "234.125782775879_9424.7314453125;",
+                                    "243.102464294434_6834.91943359375;",
+                                    "244.109477996826_1533.8046875;",
+                                    "244.146520514237_4892.56103515625;",
+                                    "368.272043863932_1573.63305664062;",
+                                    "369.206732177734_12602.3115234375;",
+                                    "370.211002349854_1442.79187011719;",
+                                    "371.222595214844_2184.82690429688;",
+                                    "491.388732910156_4496.3466796875;",
+                                    "492.324691772461_5253.52587890625"),
+            neutralprecursormolecularformula = "C10H11N4O5",
+            precursoriontype = "[M-H]-",
+            neutralprecursormass = "267.07291",
+            OfflineSpectralDatabaseFile =
+                "/vol/spectral-databases/weizfragV2.mb"
+        )
+        testCall <- POST(url = mfSrc@url,config = mfSrc@config,
+                         body = testQuery, encode = "json", handle)
+        ## a naive way to capture http 'success' codes 2xx:
+        if (length(testCall$status_code) &&
+            length(grep("^2", testCall$status_code))) mfSrc
+        else stop("Remote server inactive or irresponsive. ",
+                  "Returned status code: ", testCall$status_code)
+}
+
+.peaks_to_metfrag_string <- function(x) {
+    if (nrow(x))
+        paste0(x[, "mz"], "_", x[, "intensity"], collapse = ";")
+    else character()
+}
+
+#' @importMethodsFrom ProtGenerics as.list
+.param_to_list <- function(x, skip = c("validAdducts", "THRESHFUN")) {
+    x <- as.list(x)
+    x[!names(x) %in% skip]
 }
 
+#' @rdname MetFrag
+#'
+#' @importFrom httr GET POST content
+#'
+#' @importFrom utils read.csv
+#'
+#' @importFrom MetaboCoreUtils mz2mass
 setMethod(
 	"matchSpectra", signature(query = "Spectra", target = "MetFragSource",
                               param = "MetFragParam"),
-	function(query, target, param, BPPARAM = BiocParallel::SerialParam()) {
-        out <- list()
-		names(out) <- spectraData(query)$peak_id
-        mtch <- data.frame()	
+	function(query, target, param, debug = FALSE) {
+        validAdducts(query, param)
+        if (debug) handle <- verbose()
+        else handle <- NULL
+        out <- vector("list", length(query))
+		names(out) <- query$peak_id
+        restQuery <- .param_to_list(param)
```

Comment:
better to do this outside of the loop. no need to call this conversion for every single spectrum.